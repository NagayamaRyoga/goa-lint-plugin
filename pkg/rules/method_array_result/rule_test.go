package method_array_result_test

import (
	"log"
	"os"
	"testing"

	"github.com/NagayamaRyoga/goalint/pkg/rules/method_array_result"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"goa.design/goa/v3/dsl"
	"goa.design/goa/v3/eval"
	"goa.design/goa/v3/expr"
)

func TestRule(t *testing.T) {
	t.Parallel()

	var (
		Response = dsl.ResultType("application/vnd.response", func() {
			dsl.Required("titles")
			dsl.Attribute("titles", dsl.ArrayOf(dsl.String))
		})

		methodReturnsNonArray = dsl.Service("calc", func() {
			dsl.Method("list_titles", func() {
				dsl.Result(Response)
			})
		})

		methodReturnsArray = dsl.Service("calc2", func() {
			dsl.Method("list_titles2", func() {
				dsl.Result(dsl.ArrayOf(dsl.String))
			})
		})

		Item = dsl.ResultType("application/vnd.item", func() {
			dsl.Required("title")
			dsl.Attribute("title", dsl.String)
		})

		methodReturnsCollection = dsl.Service("calc3", func() {
			dsl.Method("list_titles3", func() {
				dsl.Result(dsl.CollectionOf(Item))
			})
		})
	)

	err := eval.RunDSL()
	assert.NoError(t, err)

	testCases := []struct {
		description string
		service     *expr.ServiceExpr
		wantReports int
	}{
		{
			description: "success",
			service:     methodReturnsNonArray,
			wantReports: 0,
		},
		{
			description: "failed/ArrayOf",
			service:     methodReturnsArray,
			wantReports: 1,
		},
		{
			description: "failed/CollectionOf",
			service:     methodReturnsCollection,
			wantReports: 1,
		},
	}
	for _, tc := range testCases {
		tc := tc
		t.Run(tc.description, func(t *testing.T) {
			t.Parallel()

			snaps := snaps.WithConfig(snaps.Dir("testdata"))

			logger := log.New(os.Stdout, "", 0)
			cfg := method_array_result.NewConfig()
			rule := method_array_result.NewRule(logger, cfg)

			// when
			assert.Equal(t, 1, len(tc.service.Methods))

			got := rule.WalkMethodExpr(tc.service.Methods[0])
			assert.Equal(t, tc.wantReports, len(got))
			snaps.MatchSnapshot(t, got.String())
		})
	}
}
