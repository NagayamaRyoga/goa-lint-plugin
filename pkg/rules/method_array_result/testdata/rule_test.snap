
[TestRule/success - 1]

---

[TestRule/failed/ArrayOf - 1]
Found 1 errors and 0 warnings
[MethodArrayResult]: error in service "calc2" method "list_titles2": Method should not return an array as the top-level structure of a response

---

[TestRule/failed/CollectionOf - 1]
Found 1 errors and 0 warnings
[MethodArrayResult]: error in service "calc3" method "list_titles3": Method should not return an array as the top-level structure of a response

---
